server/src/main/java/io/crate/execution/dml/TranslogIndexer.java

```diff
@@ -85,19 +84,29 @@ public TranslogIndexer(DocTableInfo table, Version shardCreatedVersion) {
         this.indexColumns = buildIndexColumns(
             table.indexColumns(),
             table::getReference,
-            reference -> new IndexInput(reference.column(), () -> sourceMap)
+            reference -> new IndexInput(reference.column())
         );
         ignoreUnknownColumns = table.columnPolicy() != ColumnPolicy.STRICT;
         this.shardCreatedVersion = shardCreatedVersion;
     }
 
 
-    private record IndexInput(ColumnIdent columnIdent,
-                              Supplier<Map<String, Object>> sourceSupplier) implements Input<Object> {
+    private static class IndexInput implements Input<Object> {
```

Comment:
Does this actually need to implement `Input<>`?  If we can avoid that then we don't need the unimplemented method, which feels a bit clunky.

Reply:
We won't be able to re-use Indexer's `buildIndexColumns` which is probably not a big deal (but would have to duplicate the logic). Maybe I should add the javadoc to the class explaining why we made it implement `Input` or better go with custom logic?

Reply:
Do we need the `<I extends Input<?>>` restriction in `buildIndexColumns`?

Reply:
If we want to reuse `Indexer.buildIndexColumns`, yes.
As already said, I'm not sure if its worth doing.

Reply:
No, at the end we have to use Input as essentially it all boils down to the Indexer's logic operating with inputs, ie we use expressions derived from `ctxForRefs::add `to fetch the value.

Suggestions:
1. Don't use Input at all in the TranslogIndexer and get value from the docMap directly as in https://github.com/crate/crate/pull/17403/commits/f48b17cbab0a78104541dd0eed8b4459f74eace8#diff-d08b3d5fd51f3ae7aeb3cc7d346da46d6b9aeefc68d7da5a0cbbdaf6e719fd5cR126-R132

but then ned to introduce new Map to precompute fullPath, here it was nicely encapsulated in the Input + allowed to re-use Indexer's logic (maybe not important, not sure).
2. Let it implement Input , ie leave as is, maybe add a code comment explaining why it implements Input
3. Something else?

Reply:
Right, Indexer needs to use an `Input<?>`, but `IndexColumn` itself is just a parametrized container.  So Indexer can use `IndexColumn<Input<?>>` and TranslogIndexer can use `IndexColumn<IndexInput>`; the final argument passed to `buildIndexColumn` determines the parameter type.

Reply:
In other words, I think you can replace
```
    public static <I extends Input<?>> List<IndexColumn<I>> buildIndexColumns( ...
```
with
```
public static <I> List<IndexColumn<I>> buildIndexColumns( ...
```
and everything will still work in Indexer, but IndexInput doesn't need to implement Input anymore.

Reply:
Right, thanks, pushed an update